pkg/util/assertion/assertion.go

```diff
@@ -0,0 +1,219 @@
+//	  }
+//	}
+//
+// Example: expensive assertions. Here, we assert that a range's MVCC stats are
+// accurate after every write, verifying that incremental updates done by writes
+// are correct. This is too expensive to do even in non-release builds, so we
+// gate the assertion on ExpensiveEnabled which is only enabled in special test
+// builds.
+//
+//	func (b *replicaAppBatch) ApplyToStateMachine(ctx context.Context) error {
+//	  // ...
+//	  if assertion.ExpensiveEnabled {
+//	    // runs in certain test builds
+//	    stats, err := storage.ComputeStats(b.r.store.Engine, desc.StartKey, desc.EndKey, now)
+//	    if err != nil {
+//	      return err
+//	    }
+//	    if stats != *b.state.Stats {
+//	      return assertion.Failed(ctx, "incorrect MVCC stats, %v != %v", stats, b.state.Stats)
+//	    }
+//	  }
+//	  // ...
+//	}
+//
+// Example: fatal errors, which should not be assertions. Here, Pebble detects
+// file corruption. This is not an assertion, but rather normal error handling,
+// so we fatal via log.Fatalf() even in release builds.
+//
+//	if err := iter.Close(); err == pebble.ErrCorruption {
+//	  log.Fatalf(ctx, "%v", err)
+//	}
+package assertion
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/build"
+	"github.com/cockroachdb/cockroach/pkg/util"
+	"github.com/cockroachdb/cockroach/pkg/util/envutil"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/errors"
+)
+
+var (
+	// shouldFatal exits the process with a fatal error on assertion failures. It
+	// defaults to true in non-release builds.
+	shouldFatal = envutil.EnvOrDefaultBool("COCKROACH_FATAL_ASSERTIONS", !build.IsRelease())
+
+	// MaybeSendReport is injected by package logcrash, for Sentry reporting.
+	MaybeSendReport func(ctx context.Context, err error)
+)
+
+// ExpensiveEnabled is true if expensive assertions are enabled.
+//
+// TODO(erikgrinaker): This should use a separate/different build tag, but
+// for now we gate them on race which has de-facto been used for this. See:
+// https://github.com/cockroachdb/cockroach/issues/107425
+const ExpensiveEnabled = util.RaceEnabled
+
+// Failed reports an assertion failure. In non-release builds, it fatals with a
+// stack trace. In release builds, if returns an assertion error, logs it with a
+// stack trace, and reports it to Sentry (if enabled).
+func Failed(ctx context.Context, format string, args ...interface{}) error {
+	return failedDepth(ctx, 1, format, args...)
+}
+
+// failedDepth is like Failed, but removes the given number of stack frames in
+// the stack trace and log message source code location.
+func failedDepth(ctx context.Context, depth int, format string, args ...interface{}) error {
+	depth += 1
+	err := errors.AssertionFailedWithDepthf(depth, format, args...)
+	if shouldFatal {
+		log.FatalfDepth(ctx, depth, "%+v", err)
+	} else {
+		log.ErrorfDepth(ctx, depth, "%+v", err)
+		if MaybeSendReport != nil { // can be nil in tests
+			MaybeSendReport(ctx, err)
```

Comment:
Is this blocking on the actual send? Up to 10s? Maybe call this out (to some level of detail) in higher level func comments.

A 10s delay is not a big deal if what follows is a panic/fatal, but if we keep running as normal, ideally we want this report be sent asynchronously.

Reply:
Yes, it blocks. We can't send it asynchronously, because if we fatal the process then the asynchronous code doesn't run. We could do this asynchronously for non-fatal assertions though, if we feel like it's necessary -- in that case, we'll also need to add synchronization to make sure we don't terminate the process before it's sent, and also bound the number of goroutines we spawn if we e.g. hit an assertion failure in a hot path or tight loop. For the same reason, we probably want to rate limit this.

I'll write up a follow-up issue to deal with this if/when we merge this.